### PR TITLE
feat: adding shared nullifier key pattern

### PR DESCRIPTION
**Pull request by 0xng**
# 🤖 Linear
Closes AZT-102, AZT-103

### CODE REVIEW

File: patterns/shared-nullifier-key/src/main.nr [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        get_public_key::get_public_key,
+        rand::rand,
+    };
+    use dep::aztec::state_vars::{
+        set::Set,
+    };
+
+    use dep::aztec::protocol_types::{
+        address::AztecAddress,
+    };
```
Comment (simon-something) [inline patterns/shared-nullifier-key/src/main.nr:29]: nit> just realized this: maybe tidier to group by aztec imports, wdyt?

File: patterns/shared-nullifier-key/src/main.nr [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    
+/*
+    Immediate Note Nullification
```
Comment (simon-something) [inline patterns/shared-nullifier-key/src/main.nr]: typo title

File: patterns/shared-nullifier-key/src/main.nr [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    /// @notice Consumes an SharedNote.
+    /// @param shared_nullifier The shared_nullifier used when creating the note.
+    #[aztec(private)]
+    fn nullify_note(shared_nullifier: Field) {
+        // Set the sender for readability.
+        let sender = context.msg_sender();
+
+        // Create a filter to ensure the note is not already created.
```
Comment (simon-something) [inline patterns/shared-nullifier-key/src/main.nr]: "is created" ?

File: patterns/shared-nullifier-key/src/test/e2e.test.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  
+  // Setup: Set the sandbox
+  beforeAll(async () => {
+    pxe = await setupSandbox();
+    [alice, bob, deployer] = await getInitialTestAccountsWallets(pxe);
+
+    }, 120_000);
+
+  
+  describe("E2E Shared Note", () => {
```
Comment (drgorillamd) [inline patterns/shared-nullifier-key/src/test/e2e.test.ts:38]: shared nullifier

Comment (simon-something) [inline patterns/shared-nullifier-key/src/test/e2e.test.ts]: shared nullifier instead?

File: patterns/shared-nullifier-key/src/types/shared_note.nr [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    header: NoteHeader
+}
+
+/// @notice Declare the length of the note.
+global SHARED_NOTE_LEN: Field = 2;
+
+/// @notice Declare the methods of the note.
+impl SharedNote {
+    /// @notice Creates a new note.
+    /// @param  _contex: the context. It's necessary to push the nullifier to the tree.
```
Comment (simon-something) [inline patterns/shared-nullifier-key/src/types/shared_note.nr]: typo context

**Review by simon-something — DISMISSED** [review]

File: patterns/shared-nullifier-key/src/main.nr (hunk 1 above)
Comment (0xng) [reply patterns/shared-nullifier-key/src/main.nr:29]: yes, agreed, created a task to do it at the end

**Review by simon-something — APPROVED** [review]

**Pull request merged**